### ISSUE TITLE
Supernovas: Classificação e Visualização Neural

### ISSUE THREAD

**Issue by Jacob888-dev** [ucb-bar/fpga-zynq#123]
import numpy as np
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler, LabelEncoder, OneHotEncoder
import matplotlib.pyplot as plt

# --- 1. Configurações Iniciais ---
RANDOM_SEED = 42
np.random.seed(RANDOM_SEED)
tf.random.set_seed(RANDOM_SEED)

# Parâmetros para dados sintéticos (ajuste conforme necessário)
N_SAMPLES_PER_CLASS = 300
SEQUENCE_LENGTH = 100  # Número de pontos de tempo em cada curva de luz
N_FEATURES = 1         # Por simplicidade, apenas um canal de "fluxo"
N_CLASSES = 3          # Ex: Tipo Ia, Tipo II, Tipo Ib/c

# --- 2. Geração de Dados Sintéticos ---
# (SUBSTITUA ESTA SEÇÃO PELO CARREGAMENTO E PRÉ-PROCESSAMENTO DOS SEUS DADOS REAIS)
def generate_synthetic_light_curve(class_type, length=SEQUENCE_LENGTH):
    time = np.linspace(0, length-1, length)
    flux = np.zeros(length)
    peak_time = length / 2

    if class_type == 0: # "Tipo Ia" - subida rápida, declínio exponencial
        rise_length = int(peak_time / 2)
        flux[:rise_length] = np.linspace(0, 1, rise_length)**2
        flux[rise_length:length] = np.exp(-(time[rise_length:length] - time[rise_length]) / (length / 4))
        flux[rise_length:] *= 1.0 # Peak
    elif class_type == 1: # "Tipo II-P" - platô
        rise_length = int(peak_time / 3)
        plateau_start = rise_length
        plateau_end = int(length * 0.7)
        flux[:rise_length] = np.linspace(0, 0.8, rise_length)**1.5
        flux[plateau_start:plateau_end] = 0.8 + np.random.normal(0, 0.05, plateau_end - plateau_start)
        flux[plateau_end:length] = 0.8 * np.exp(-(time[plateau_end:length] - time[plateau_end]) / (length / 8))
    else: # "Tipo Ib/c" - sem hidrogênio, declínio mais rápido que Ia após o pico
        rise_length = int(peak_time / 2.5)
        flux[:rise_length] = np.linspace(0, 0.9, rise_length)**1.8
        flux[rise_length:length] = 0.9 * np.exp(-(time[rise_length:length] - time[rise_length]) / (length / 5))

    # Adicionar um pouco de ruído
    flux += np.random.normal(0, 0.05, length)
    return flux.reshape(-1, 1) # (SEQUENCE_LENGTH, N_FEATURES)

light_curves = []
labels = []

for i in range(N_CLASSES):
    for _ in range(N_SAMPLES_PER_CLASS):
        light_curves.append(generate_synthetic_light_curve(i))
        labels.append(i)

X = np.array(light_curves)
y = np.array(labels)

print(f"Forma dos dados X: {X.shape}") # (N_SAMPLES_TOTAL, SEQUENCE_LENGTH, N_FEATURES)
print(f"Forma dos rótulos y: {y.shape}")

# Visualizar alguns exemplos de dados sintéticos
plt.figure(figsize=(12, 6))
for i in range(N_CLASSES):
    idx = np.where(y == i)[0][0]
    plt.plot(X[idx, :, 0], label=f'Classe Sintética {i}')
plt.title("Exemplos de Curvas de Luz Sintéticas")
plt.xlabel("Tempo (dias simulados)")
plt.ylabel("Fluxo Normalizado (simulado)")
plt.legend()
plt.show()

# --- 3. Pré-processamento de Dados ---
# Normalização (MinMax para cada curva individualmente ou globalmente)
# Para este exemplo, normalizaremos globalmente após a divisão para evitar data leakage,
# mas na prática, para curvas de luz, pode ser melhor normalizar cada curva por seu pico.
# Aqui, como os dados sintéticos já estão mais ou menos na faixa [0,1], vamos simplificar.
# Se seus fluxos tiverem escalas muito diferentes, a normalização é crucial.
# scaler = MinMaxScaler()
# X_scaled = []
# for lc in X:
#     X_scaled.append(scaler.fit_transform(lc)) # Normaliza cada curva individualmente
# X = np.array(X_scaled)

# Codificação dos Rótulos
label_encoder = LabelEncoder()
y_encoded = label_encoder.fit_transform(y)
onehot_encoder = OneHotEncoder(sparse_output=False) # sparse=False for TF 2.x compatibility
y_onehot = onehot_encoder.fit_transform(y_encoded.reshape(-1, 1))

# Divisão em Treino, Validação e Teste
X_train, X_temp, y_train, y_temp = train_test_split(X, y_onehot, test_size=0.3, random_state=RANDOM_SEED, stratify=y_onehot)
X_val, X_test, y_val, y_test = train_test_split(X_temp, y_temp, test_size=0.5, random_state=RANDOM_SEED, stratify=y_temp)

print(f"X_train shape: {X_train.shape}, y_train shape: {y_train.shape}")
print(f"X_val shape: {X_val.shape}, y_val shape: {y_val.shape}")
print(f"X_test shape: {X_test.shape}, y_test shape: {y_test.shape}")

# --- 4. Construção do Modelo de Classificação ---
def build_classifier_model(input_shape, num_classes):
    model = keras.Sequential(
        [
            layers.Input(shape=input_shape),
            # Camadas Convolucionais 1D para extrair features locais da sequência
            layers.Conv1D(filters=64, kernel_size=7, activation="relu", padding="same"),
            layers.BatchNormalization(),
            layers.MaxPooling1D(pool_size=2),
            layers.Conv1D(filters=128, kernel_size=5, activation="relu", padding="same"),
            layers.BatchNormalization(),
            layers.MaxPooling1D(pool_size=2),
            # Camada LSTM para processar a sequência de features
            layers.LSTM(128, return_sequences=False), # return_sequences=False se a próxima camada for Densa
            layers.Dropout(0.4),
            # Camadas Densas para classificação
            layers.Dense(64, activation="relu"),
            layers.Dropout(0.3),
            layers.Dense(num_classes, activation="softmax"),
        ]
    )
    return model

input_shape = (SEQUENCE_LENGTH, N_FEATURES)
classifier_model = build_classifier_model(input_shape, N_CLASSES)
classifier_model.summary()

# --- 5. Treinamento do Modelo ---
EPOCHS = 50 # Aumente para dados reais
BATCH_SIZE = 32

classifier_model.compile(
    optimizer=keras.optimizers.Adam(learning_rate=1e-4),
    loss="categorical_crossentropy",
    metrics=["accuracy"],
)

# Callbacks
early_stopping = keras.callbacks.EarlyStopping(monitor='val_loss', patience=10, restore_best_weights=True)
reduce_lr = keras.callbacks.ReduceLROnPlateau(monitor='val_loss', factor=0.2, patience=5, min_lr=1e-6)

history = classifier_model.fit(
    X_train,
    y_train,
    epochs=EPOCHS,
    batch_size=BATCH_SIZE,
    validation_data=(X_val, y_val),
    callbacks=[early_stopping, reduce_lr],
    verbose=1
)

# --- 6. Avaliação do Modelo ---
loss, accuracy = classifier_model.evaluate(X_test, y_test, verbose=0)
print(f"\nResultados no conjunto de Teste:")
print(f"Perda (Loss): {loss:.4f}")
print(f"Acurácia: {accuracy*100:.2f}%")

# Plotar histórico de treinamento
plt.figure(figsize=(12, 4))
plt.subplot(1, 2, 1)
plt.plot(history.history['accuracy'], label='Acurácia Treino')
plt.plot(history.history['val_accuracy'], label='Acurácia Validação')
plt.title('Acurácia do Modelo')
plt.xlabel('Época')
plt.ylabel('Acurácia')
plt.legend()

plt.subplot(1, 2, 2)
plt.plot(history.history['loss'], label='Perda Treino')
plt.plot(history.history['val_loss'], label='Perda Validação')
plt.title('Perda do Modelo')
plt.xlabel('Época')
plt.ylabel('Perda')
plt.legend()
plt.tight_layout()
plt.show()

# --- 7. "Shader Neural" Conceitual: Visualização Estilizada ---
# Esta é uma DEMONSTRAÇÃO MUITO SIMPLES.
# Um "shader neural" real envolveria outra rede neural para gerar/transformar a visualização.

def visualize_light_curve_with_style(light_curve_data, true_label_idx, predicted_label_idx, class_names=None):
    """
    Plota a curva de luz e aplica um estilo simples baseado na predição.
    """
    if class_names is None:
        class_names = [f"Tipo {i}" for i in range(N_CLASSES)]

    true_class_name = class_names[true_label_idx]
    predicted_class_name = class_names[predicted_label_idx]

    plt.figure(figsize=(8, 5))
    plt.plot(light_curve_data[:, 0], label=f"Curva de Luz (Fluxo)", color='gray', alpha=0.8)

    # Estilização baseada na predição
    color = 'green' # Cor padrão para predição correta
    title_info = f"Previsto: {predicted_class_name} (Verdadeiro: {true_class_name})"

    if predicted_label_idx != true_label_idx:
        color = 'red' # Cor para predição incorreta
        title_info += " - INCORRETO"

    # Simular um "efeito" do shader neural mudando a cor do título ou da linha de pico
    # Aqui, apenas mudamos a cor da anotação da classe
    plt.text(0.05, 0.95, f"Pred: {predicted_class_name}",
             transform=plt.gca().transAxes, fontsize=12,
             verticalalignment='top', color=color,
             bbox=dict(boxstyle='round,pad=0.3', fc='white', alpha=0.7, ec=color))

    plt.title(title_info)
    plt.xlabel("Tempo (dias simulados)")
    plt.ylabel("Fluxo Normalizado (simulado)")
    plt.legend(loc='lower right')
    plt.grid(True, linestyle='--', alpha=0.6)
    plt.show()

# Fazer previsões no conjunto de teste
y_pred_probs = classifier_model.predict(X_test)
y_pred_indices = np.argmax(y_pred_probs, axis=1)
y_true_indices = np.argmax(y_test, axis=1) # Converter de one-hot para índice

# Visualizar algumas curvas de luz do conjunto de teste com o "shader" conceitual
print("\nVisualizando algumas predições com estilo 'shader neural' conceitual:")
num_visualize = 5
for i in range(num_visualize):
    idx = np.random.randint(0, len(X_test)) # Pegar um índice aleatório
    visualize_light_curve_with_style(
        X_test[idx],
        y_true_indices[idx],
        y_pred_indices[idx],
        class_names=[f"SN Tipo {chr(65+i)}" for i in range(N_CLASSES)] # Nomes A, B, C...
    )

# --- Considerações para um "Shader Neural" mais avançado ---
print("""
Considerações para um "Shader Neural" mais avançado (além do exemplo de plotagem):
1.  Redes Gerativas (GANs, VAEs):
    * Treinar uma GAN para gerar imagens estilizadas de explosões condicionadas à curva de luz ou à classe prevista.
    * O gerador da GAN atuaria como o "shader neural".
    * Exigiria um conjunto de dados de pares (curva de luz/parâmetros -> imagem/animação da explosão).
2.  Transferência de Estilo Neural:
    * Aplicar o estilo de uma imagem artística (ex: nebulosa, pintura abstrata) a uma representação base da supernova.
    * O "shader" seria a rede de transferência de estilo.
3.  NeRF (Neural Radiance Fields):
    * Se dados multi-vista de simulações de explosões existirem, um NeRF poderia aprender uma representação 3D.
    * O "shader" seria o processo de renderização do NeRF.
4.  Entradas para o Shader Neural:
    * Curva de luz bruta.
    * Representações latentes da rede de classificação.
    * Parâmetros físicos inferidos (energia, massa ejetada).
    * Classe prevista.
5.  Saídas do Shader Neural:
    * Imagens estáticas.
    * Animações.
    * Visualizações interativas (se integrado a um motor gráfico).

Este código fornece a base para a classificação. O 'shader neural' é um campo ativo de pesquisa
e exigiria ferramentas e técnicas mais especializadas em computação gráfica e IA generativa.
""")
